File: claude/Semester-4/B-FUN-400/Wolfram/Wolfram_roro/src/Wolfram.hs
{-
EPITECH PROJECT, 2025
Wolfram
File description:
Main Wolfram module for cellular automata
-}

module Wolfram
    ( CellState(..)
    , Generation
    , Rule
    , Config(..)
    , applyRule
    , nextGeneration
    , generateInitialState
    , displayGeneration
    , runAutomaton
    ) where

import System.IO (hFlush, stdout)

data CellState = Dead | Alive
    deriving (Eq, Show)

type Generation = [CellState]
type Rule = (CellState, CellState, CellState) -> CellState

data Config = Config
    { configRule    :: Int
    , configStart   :: Int
    , configLines   :: Maybe Int
    , configWindow  :: Int
    , configMove    :: Int
    } deriving (Show)

-- Rule definitions for cellular automata
ruleTable :: Int -> Rule
ruleTable n = \(left, center, right) -> 
    let index = cellToInt left * 4 + cellToInt center * 2 + cellToInt right
        bit = (n `div` (2 ^ index)) `mod` 2
    in if bit == 1 then Alive else Dead
  where
    cellToInt Dead = 0
    cellToInt Alive = 1

-- Apply rule to generate next cell state
applyRule :: Rule -> Generation -> Int -> CellState
applyRule rule gen pos = rule (getCell (pos - 1), getCell pos, getCell (pos + 1))
  where
    getCell i
        | i < 0 || i >= length gen = Dead
        | otherwise = gen !! i

-- Generate next generation from current one
nextGeneration :: Rule -> Generation -> Generation
nextGeneration rule gen = 
    let len = length gen
        extended = replicate 100 Dead ++ gen ++ replicate 100 Dead
        newGen = map (applyRule rule extended) [100..len + 99]
    in newGen

-- Generate initial state with single alive cell in center
generateInitialState :: Int -> Generation
generateInitialState width = 
    let center = width `div` 2
    in replicate center Dead ++ [Alive] ++ replicate (width - center - 1) Dead

-- Display a generation as a string
displayGeneration :: Generation -> Int -> Int -> String
displayGeneration gen window move =
    let startPos = max 0 (length gen `div` 2 - window `div` 2 + move)
        endPos = min (length gen) (startPos + window)
        visibleCells = take (endPos - startPos) $ drop startPos gen
        cellToChar Dead = ' '
        cellToChar Alive = '*'
    in map cellToChar visibleCells

-- Run the automaton simulation
runAutomaton :: Config -> IO ()
runAutomaton config = do
    let rule = ruleTable (configRule config)
        initial = generateInitialState 200
        generations = iterate (nextGeneration rule) initial
        startGen = configStart config
        linesToShow = maybe maxBound id (configLines config)
    
    mapM_ (\(i, gen) -> do
        putStrLn $ displayGeneration gen (configWindow config) (configMove config)
        hFlush stdout
        ) $ take linesToShow $ zip [startGen..] $ drop startGen generations

File: claude/Semester-4/B-FUN-400/Wolfram/Wolfram_roro/wolfram.py
#!/usr/bin/env python3
"""
EPITECH PROJECT, 2025
Wolfram
File description:
Elementary Cellular Automaton implementation
"""

import sys
import argparse

def rule_table(rule_number):
    """Create rule table for given rule number"""
    table = {}
    for i in range(8):
        left = (i >> 2) & 1
        center = (i >> 1) & 1
        right = i & 1
        bit = (rule_number >> i) & 1
        table[(left, center, right)] = bit
    return table

def next_generation(current_gen, rule_table):
    """Generate next generation from current one"""
    width = len(current_gen)
    extended = [0] * 100 + current_gen + [0] * 100
    new_gen = []
    
    for i in range(100, width + 100):
        left = extended[i-1] if i-1 >= 0 else 0
        center = extended[i]
        right = extended[i+1] if i+1 < len(extended) else 0
        new_gen.append(rule_table.get((left, center, right), 0))
    
    return new_gen

def display_generation(generation, window, move):
    """Display a generation as a string"""
    width = len(generation)
    center = width // 2
    start_pos = max(0, center - window // 2 + move)
    end_pos = min(width, start_pos + window)
    
    visible_cells = generation[start_pos:end_pos]
    return ''.join('*' if cell else ' ' for cell in visible_cells)

def show_usage():
    """Display usage message"""
    print("USAGE: wolfram --rule R [--start N] [--lines N] [--window N] [--move N]", file=sys.stderr)
    print("", file=sys.stderr)
    print("    --rule      the ruleset to use (mandatory)", file=sys.stderr)
    print("    --start     the generation number at which to start the display (default: 0)", file=sys.stderr)
    print("    --lines     the number of lines to display (default: infinite)", file=sys.stderr)
    print("    --window    the number of cells to display on each line (default: 80)", file=sys.stderr)
    print("    --move      translation to apply on the window (default: 0)", file=sys.stderr)

def main():
    """Main function"""
    try:
        # Manual argument parsing (getopt is forbidden)
        args = sys.argv[1:]
        if not args:
            show_usage()
            sys.exit(84)
        
        # Default values
        rule_num = None
        start = 0
        lines = None
        window = 80
        move = 0
        
        i = 0
        while i < len(args):
            if args[i] == '--rule':
                if i + 1 >= len(args):
                    show_usage()
                    sys.exit(84)
                try:
                    rule_num = int(args[i + 1])
                    if not (0 <= rule_num <= 255):
                        show_usage()
                        sys.exit(84)
                except ValueError:
                    show_usage()
                    sys.exit(84)
                i += 2
            elif args[i] == '--start':
                if i + 1 >= len(args):
                    show_usage()
                    sys.exit(84)
                try:
                    start = int(args[i + 1])
                    if start < 0:
                        show_usage()
                        sys.exit(84)
                except ValueError:
                    show_usage()
                    sys.exit(84)
                i += 2
            elif args[i] == '--lines':
                if i + 1 >= len(args):
                    show_usage()
                    sys.exit(84)
                try:
                    lines = int(args[i + 1])
                    if lines <= 0:
                        show_usage()
                        sys.exit(84)
                except ValueError:
                    show_usage()
                    sys.exit(84)
                i += 2
            elif args[i] == '--window':
                if i + 1 >= len(args):
                    show_usage()
                    sys.exit(84)
                try:
                    window = int(args[i + 1])
                    if window <= 0:
                        show_usage()
                        sys.exit(84)
                except ValueError:
                    show_usage()
                    sys.exit(84)
                i += 2
            elif args[i] == '--move':
                if i + 1 >= len(args):
                    show_usage()
                    sys.exit(84)
                try:
                    move = int(args[i + 1])
                except ValueError:
                    show_usage()
                    sys.exit(84)
                i += 2
            else:
                show_usage()
                sys.exit(84)
        
        if rule_num is None:
            show_usage()
            sys.exit(84)
        
        # Create initial generation with single alive cell in center
        initial_width = 200
        initial_gen = [0] * (initial_width // 2) + [1] + [0] * (initial_width // 2)
        
        # Create rule table
        rule = rule_table(rule_num)
        
        # Generate and display automaton
        current_gen = initial_gen
        gen_count = 0
        lines_displayed = 0
        
        # Skip generations until start
        for _ in range(start):
            current_gen = next_generation(current_gen, rule)
        
        # Display generations
        while lines is None or lines_displayed < lines:
            print(display_generation(current_gen, window, move))
            sys.stdout.flush()
            current_gen = next_generation(current_gen, rule)
            lines_displayed += 1
        
    except Exception:
        sys.exit(84)

if __name__ == "__main__":
    main()